File: ports/go/security/locale_case_inversion.go
package security

import "slices"

// Locale-case-inversion detector (UAX #21 / Tier A2) — inputs whose lowercase
// fold inverts across locales, the homograph-via-locale attack (CVE-2007-6692,
// CVE-2021-30245, the Spotify "İSTANBUL" / "iSTANBUL" incident class). Mirrors
// Unicode.Security.Form.LocaleCaseInversion.
//
// Detection compares per-position lowerCodepoint under each locale against the
// default, rather than diffing whole-string toLower, because lowerCodepoint
// evaluates the SpecialCasing context predicates with full surrounding context.
// Turkish divergence takes priority over Lithuanian (SpecialCasing has no
// az-only codepoint, so Turkish covers Azeri).

// lowerCodepoint lowercases a single codepoint in its full input context: the
// SpecialCasing row whose conditions hold, else the simple lowercase mapping.
func lowerCodepoint(locale casingLocale, revPrefix, suffix []uint32, cp uint32) []uint32 {
	if row, ok := findSpecialRow(locale, revPrefix, suffix, cp); ok {
		return row.lower
	}
	return []uint32{simpleLowercase(cp)}
}

// firstLocaleDivergence returns the first input position whose lowerCodepoint
// under locale differs from the default-locale result.
func firstLocaleDivergence(locale casingLocale, input []uint32) (int, bool) {
	revPrefix := []uint32{}
	for i, cp := range input {
		suffix := input[i+1:]
		defaultLower := lowerCodepoint(localeDefault, revPrefix, suffix, cp)
		localeLower := lowerCodepoint(locale, revPrefix, suffix, cp)
		if !slices.Equal(defaultLower, localeLower) {
			return i, true
		}
		revPrefix = append([]uint32{cp}, revPrefix...)
	}
	return 0, false
}

// localeCaseInversionDetect reports the divergent-locale sub-threat and the
// first divergent position. Turkish divergence takes priority; Lithuanian is
// reached only when no Turkish divergence is found.
func localeCaseInversionDetect(input []uint32) (string, []int, bool) {
	if pos, ok := firstLocaleDivergence(localeTurkish, input); ok {
		return "TurkishCaseDivergence", []int{pos}, true
	}
	if pos, ok := firstLocaleDivergence(localeLithuanian, input); ok {
		return "LithuanianCaseDivergence", []int{pos}, true
	}
	return "", nil, false
}

File: ports/ruby/lib/unicode_ruby/security/form/locale_case_inversion.rb
# frozen_string_literal: true

require_relative "../identity/ucd"

module UnicodeRuby
  module Security
    module Form
      # Locale-case-inversion — inputs whose lowercase result inverts across
      # locales (CVE-2007-6692, CVE-2021-30245, the Spotify "İSTANBUL" class).
      # Compares per-position lower_codepoint under each locale against the
      # default; Turkish divergence (which covers Azeri) takes priority over
      # Lithuanian.
      module LocaleCaseInversion
        Detection = Struct.new(:sub, :positions)

        module_function

        # First input position whose lower_codepoint under `locale` differs from
        # the default-locale result, with the codepoint at that position.
        def first_locale_divergence(locale, input)
          rev_prefix = []
          input.each_with_index do |cp, index|
            suffix = input[(index + 1)..]
            default_lower = Ucd.lower_codepoint(Ucd::Locale::DEFAULT, rev_prefix, suffix, cp)
            locale_lower = Ucd.lower_codepoint(locale, rev_prefix, suffix, cp)
            return [index, cp] if default_lower != locale_lower

            rev_prefix.unshift(cp)
          end
          nil
        end

        def detect(input)
          turkish = first_locale_divergence(Ucd::Locale::TURKISH, input)
          return Detection.new("TurkishCaseDivergence", [turkish[0]]) unless turkish.nil?

          lithuanian = first_locale_divergence(Ucd::Locale::LITHUANIAN, input)
          return Detection.new("LithuanianCaseDivergence", [lithuanian[0]]) unless lithuanian.nil?

          Detection.new(nil, [])
        end
      end
    end
  end
end
